### ISSUE TITLE
Starup issue, file system provider and new workspace– no provider for memfs:/

### ISSUE THREAD

**Issue by mhchem** [Microsoft/vscode#57989]
This is similar to #51088, but the fix there does not fix the issue here.

- Install Sample file system provider extension from https://github.com/Microsoft/vscode-extension-samples/tree/master/fsprovider-sample
- In `package.json`, change the activiation to `*`, so that it gets loaded on startup.
```
"activationEvents": [
    "*"
],
```
- In `extension.ts` comment the lines 58, 60, so that `memfs.workspaceInit` will be executed within `activate()`
- run

The error message `no provider for memfs:/` will show up.

Basically, this calls `registerFileSystemProvider('memfs'...` and `updateWorkspaceFolders(0, 0, { uri: vscode.Uri.parse('memfs:/')...` in the same function.

How can I defer the second function until the file system provider is ready? Or what other recommended way is there for an extension that add a new workspace folder with the file system it provides itself?

(Everything is very complicated, because the second command restarts the whole extension. Fortunately, an infinite loop is prevented.)

Win 10, v1.26.1

**Comment by jrieken** [comment]
@alexandrudima  * activation does seem to wait for activate calls to finish...

**Comment by alexandrudima** [comment]
@jrieken Can you rephrase? What do you mean?

**Comment by alexandrudima** [comment]
`*` activation is not necessarily **stronger** than e.g. `onFileSystem:`. This is a distinction we've made since day 1 and is intentional. `*` activation means that your extension gets loaded on startup, but it does not necessarily mean that all VS Code operations will be blocked until your extension is activated.

For example, if we have an activation event `onFileSystem:memfs` and we send that out, that will not wait for all `*` extensions to finish activating. So, `*` means your extension gets activated on startup but it does not mean that it will cause any piece of VSCode to **wait** for your activation. This is intentional, because there are a lot of extensions that decide to activate on startup, but we don't want to wait for 20 startup activating extensions to be able to type, start debugging, etc.

You can use `activationEvents: [ "*", "onFileSystem:memfs" ]` to activate on startup and let vscode know that you are providing a file system for the scheme `memfs` and that your activate call should be waited by the file system handling logic of vscode.

**Comment by jrieken** [comment]
a `not` was missing: an extension activates on `*`, we call the `activate` function, we don't wait for it to finish

**Comment by jrieken** [comment]
sounds like `as-designed`

**Comment by alexandrudima** [comment]
@jrieken Yes, that is how it works. 

[This](https://github.com/Microsoft/vscode/blob/94c4825153bbf06434a832e3a22404801236e916/src/vs/workbench/services/files/electron-browser/remoteFileService.ts#L257) means that the promise will wait for the activation of extensions that use `onFileSystem:X` and not for the activation of extensions that use `onFileSystem:X` or `*`.

If you want to make file system providers wait for `*` activated extensions, that is possible by adding the explicit call to `activateByEvent('*')`. But I would not do that, since so many extensions are activated on `*`.

I can look into improving our documentation in this area.

**Comment by mhchem** [comment]
Thanks, I understand.

@alexandrudima The 2nd line read "that use `onFileSystem:Y` or `*`" (`Y`, not `X`), right?

------

I put my findings here, because they answer the initial posting and might help others.

When creating an extension that provides a file system and adds corresponding workspace folder on startup, consider the following.
Make sure to have the `activationEvent` `onFileSystem:memfs` noted, otherwise you will get an `no provider for memfs:/` error, because `updateWorkspaceFolders` re-activates the extension and needs to know to wait. BTW, `onFileSystem:memfs` works side-by-side with `*`.
If you add the workspace folder from within `activate()`, you have to guard it with something like
```
if (undefined === vscode.workspace.getWorkspaceFolder(vscode.Uri.parse('memfs:/'))) {
    vscode.workspace.updateWorkspaceFolders(0, 0, { uri: vscode.Uri.parse('memfs:/'), name: "MemFS - Sample" });
}
```
otherwise you will run into a recursion (because `updateWorkspaceFolders` re-activates the extension), most probably indicated by an `Error: write EPIPE`.
